File: go/LeetCode/rotateRight.go
package main

func main() {
}

func rotateRight(head *ListNode, k int) *ListNode {
	if k==0 || head ==nil || head.Next == nil {
		return head
	}
	count:=1
	temp := head
	for temp.Next!=nil {
		temp = temp.Next
		count++
	}

	len:= count - k%count
	if len == count {
		return head
	}

	temp.Next = head

	for len>0 {
		temp=temp.Next
		len--
	}

	res := temp.Next
	temp.Next=nil
	return res
}

func rotateRight1(head *ListNode, k int) *ListNode {
	if k==0||head== nil||head.Next==nil{
		return head
	}

	count:=0
	temp:=head
	for temp != nil {
		temp = temp.Next
		count++
	}

	k = k%count
	fast := head
	slow := head

	for k>0 {
		fast = fast.Next
		k--
	}

	for fast.Next!=nil {
		fast=fast.Next
		slow=slow.Next
	}
	fast.Next = head
	head = slow.Next
	slow.Next=nil
	return head
}

File: python3/LeetCode/rotateRight.py
from typing import Optional

from ListNode import ListNode


class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k==0 or head is None or head.next is None:
            return head

        count = 1
        temp = head
        while not temp.next is None:
            temp = temp.next
            count+=1

        len = count-k%count
        if len == count:
            return head

        temp.next = head

        while len>0:
            temp = temp.next
            len-=1

        res = temp.next
        temp.next = None
        return res

    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k==0 or head is None or head.next is None:
            return head

        count =0
        temp = head
        while not temp is None:
            temp = temp.next
            count+=1

        k=k%count
        fast = head
        slow = head

        while k>0:
            fast = fast.next
            k-=1

        while not fast.next is None:
            fast = fast.next
            slow = slow.next

        fast.next = head
        head = slow.next
        slow.next = None
        return head

